File: backend/app/services/course_service.py
from typing import List, Optional
from sqlalchemy.orm import Session
from fastapi import HTTPException, status

from .. import models, schemas
from ..models.models import Module  # Import direct du modèle Module

class CourseService:
    @staticmethod
    def get_course(db: Session, course_id: int) -> models.Course:
        """Récupère un cours par son ID avec une structure cohérente"""
        from sqlalchemy.orm import joinedload
        
        # Charger le cours avec toutes les relations nécessaires
        db_course = db.query(models.Course).options(
            joinedload(models.Course.modules)
            .joinedload(models.Module.lessons),
            joinedload(models.Course.category),
            joinedload(models.Course.tags)
        ).filter(models.Course.id == course_id).first()
        
        if not db_course:
            raise HTTPException(status_code=404, detail="Cours non trouvé")
        
        # Si le cours a des leçons directes et aucun module, créer un module par défaut
        if hasattr(db_course, 'lessons') and db_course.lessons and not db_course.modules:
            from datetime import datetime
            
            # Créer un module par défaut
            default_module = models.Module(
                title="Contenu du cours",
                description="Toutes les leçons de ce cours",
                order_index=1,
                course_id=db_course.id,
                created_at=datetime.utcnow(),
                updated_at=datetime.utcnow()
            )
            
            # Déplacer les leçons vers le module
            for lesson in db_course.lessons:
                lesson.module = default_module
            
            # Ajouter le module au cours
            if not hasattr(db_course, 'modules') or db_course.modules is None:
                db_course.modules = []
                
            db_course.modules.append(default_module)
            
            # Nettoyer les leçons directes pour éviter la redondance
            db_course.lessons = []
            
            # Sauvegarder les changements
            db.add(default_module)
            db.commit()
            db.refresh(db_course)
        
        return db_course
    
    @staticmethod
    def get_courses(
        db: Session, 
        skip: int = 0, 
        limit: int = 100,
        level: Optional[str] = None,
        tag: Optional[str] = None,
        search: Optional[str] = None,
        user_id: Optional[int] = None
    ) -> List[models.Course]:
        """Récupère une liste de cours avec filtres optionnels et progression utilisateur"""
        from sqlalchemy.orm import joinedload, contains_eager
        from sqlalchemy import or_
        
        print(f"\n--- SERVICE COURSE ---")
        print(f"Paramètres reçus: skip={skip}, limit={limit}, level={level}, tag={tag}, search={search}, user_id={user_id}")
        
        # Requête de base pour les cours
        query = db.query(models.Course)
        print(f"Requête de base créée")
        
        # Chargement des relations
        query = query.options(
            joinedload(models.Course.category),
            joinedload(models.Course.tags),
            joinedload(models.Course.modules)
            .joinedload(models.Module.lessons)
            .joinedload(models.Lesson.completed_by)
        )
        
        # Filtrage par niveau
        if level:
            query = query.filter(models.Course.level == level)
            
        # Filtrage par tag
        if tag:
            query = query.join(models.Course.tags).filter(models.Tag.name == tag)
            
        # Recherche par mot-clé
        if search:
            search = f"%{search}%"
            query = query.filter(
                (models.Course.title.ilike(search)) | 
                (models.Course.description.ilike(search))
            )
        
        # Exécution de la requête
        print(f"Exécution de la requête avec offset={skip}, limit={limit}")
        courses = query.offset(skip).limit(limit).all()
        print(f"Nombre de cours récupérés de la DB: {len(courses)}")
        
        if courses:
            print(f"Cours trouvés: {[(c.id, c.title) for c in courses[:3]]}")
        else:
            print("Aucun cours trouvé dans la base de données")
            # Vérifions s'il y a des cours du tout
            total_courses = db.query(models.Course).count()
            print(f"Nombre total de cours dans la DB: {total_courses}")
        
        # Si un user_id est fourni, marquer les leçons complétées
        if user_id is not None:
            print(f"Marquage des leçons complétées pour l'utilisateur {user_id}")
            for course in courses:
                for module in course.modules:
                    for lesson in module.lessons:
                        # Vérifier si la leçon est complétée par l'utilisateur
                        lesson.is_completed = any(
                            completion.user_id == user_id 
                            for completion in lesson.completed_by
                        )
        
        print(f"--- FIN SERVICE COURSE ---\n")
        return courses
    
    @staticmethod
    def create_course(db: Session, course: schemas.CourseCreate) -> models.Course:
        """Crée un nouveau cours"""
        # Vérifie si un cours avec le même titre existe déjà
        db_course = db.query(models.Course).filter(
            models.Course.title == course.title
        ).first()
        
        if db_course:
            raise HTTPException(
                status_code=400,
                detail="Un cours avec ce titre existe déjà"
            )
        
        # Crée le cours
        db_course = models.Course(
            title=course.title,
            description=course.description,
            thumbnail_url=course.thumbnail_url,
            level=course.level,
            duration=course.duration,
            is_published=course.is_published
        )
        
        # Ajoute les tags
        for tag_name in course.tags:
            # Vérifie si le tag existe déjà
            db_tag = db.query(models.Tag).filter(
                models.Tag.name == tag_name
            ).first()
            
            if not db_tag:
                # Crée un nouveau tag s'il n'existe pas
                db_tag = models.Tag(name=tag_name)
                db.add(db_tag)
                db.commit()
                db.refresh(db_tag)
            
            db_course.tags.append(db_tag)
        
        db.add(db_course)
        db.commit()
        db.refresh(db_course)
        
        return db_course
    
    @staticmethod
    def update_course(
        db: Session, 
        course_id: int, 
        course_update: schemas.CourseUpdate
    ) -> models.Course:
        """Met à jour un cours existant"""
        db_course = CourseService.get_course(db, course_id)
        
        # Met à jour les champs fournis
        update_data = course_update.dict(exclude_unset=True)
        for field, value in update_data.items():
            setattr(db_course, field, value)
        
        db.commit()
        db.refresh(db_course)
        return db_course
    
    @staticmethod
    def delete_course(db: Session, course_id: int) -> bool:
        """Supprime un cours"""
        db_course = CourseService.get_course(db, course_id)
        db.delete(db_course)
        db.commit()
        return True
    
    @staticmethod
    def create_module(
        db: Session, 
        course_id: int, 
        module: schemas.ModuleCreate
    ) -> Module:
        """Crée un nouveau module pour un cours"""
        # Vérifie si le cours existe
        db_course = CourseService.get_course(db, course_id)
        
        # Vérifie si un module avec le même ordre existe déjà
        db_module = db.query(Module).filter(
            Module.course_id == course_id,
            Module.order_index == module.order
        ).first()
        
        if db_module:
            raise HTTPException(
                status_code=400,
                detail="Un module avec cet ordre existe déjà pour ce cours"
            )
        
        # Crée le module
        db_module = Module(
            title=module.title,
            description=module.description,
            order=module.order,
            course_id=course_id
        )
        
        db.add(db_module)
        db.commit()
        db.refresh(db_module)
        
        return db_module
    
    @staticmethod
    def create_lesson(
        db: Session, 
        module_id: int, 
        lesson: schemas.LessonCreate
    ) -> models.Lesson:
        """Crée une nouvelle leçon pour un module"""
        # Vérifie si le module existe
        db_module = db.query(models.Module).filter(
            models.Module.id == module_id
        ).first()
        
        if not db_module:
            raise HTTPException(status_code=404, detail="Module non trouvé")
        
        # Vérifie si une leçon avec le même ordre existe déjà
        db_lesson = db.query(models.Lesson).filter(
            models.Lesson.module_id == module_id,
            models.Lesson.order == lesson.order
        ).first()
        
        if db_lesson:
            raise HTTPException(
                status_code=400,
                detail="Une leçon avec cet ordre existe déjà pour ce module"
            )
        
        # Crée la leçon
        db_lesson = models.Lesson(
            title=lesson.title,
            description=lesson.description,
            content=lesson.content,
            video_url=lesson.video_url,
            duration=lesson.duration,
            order=lesson.order,
            module_id=module_id
        )
        
        db.add(db_lesson)
        db.commit()
        db.refresh(db_lesson)
        
        return db_lesson
    
    @staticmethod
    def get_course_with_progress(
        db: Session, 
        course_id: int, 
        user_id: int
    ) -> schemas.CourseWithProgress:
        """Récupère un cours avec les informations de progression de l'utilisateur"""
        from sqlalchemy.orm import joinedload
        
        # Récupère le cours avec une structure normalisée
        db_course = CourseService.get_course(db, course_id)
        
        # Vérifie si l'utilisateur est inscrit au cours
        user_progress = db.query(models.UserProgress).filter(
            models.UserProgress.user_id == user_id,
            models.UserProgress.course_id == course_id
        ).first()
        
        # Calcule la progression globale du cours
        progress = 0.0
        is_enrolled = user_progress is not None
        
        if is_enrolled and user_progress:
            progress = user_progress.completion_percentage
        
        # Marquer les leçons complétées par l'utilisateur
        if is_enrolled and hasattr(db_course, 'modules'):
            completed_lessons = db.query(models.LessonCompletion.lesson_id).filter(
                models.LessonCompletion.user_id == user_id
            ).all()
            completed_lesson_ids = {lesson_id for (lesson_id,) in completed_lessons}
            
            for module in db_course.modules:
                for lesson in module.lessons:
                    lesson.is_completed = lesson.id in completed_lesson_ids
        
        # Convertit le modèle en Pydantic avec les informations de progression
        course_dict = db_course.__dict__
        course_dict["progress"] = progress
        course_dict["is_enrolled"] = is_enrolled
        
        return schemas.CourseWithProgress(**course_dict)

File: frontend/src/services/course.service.ts
import { apiService } from './api';
import API_CONFIG from '@/config/api';
import type { 
  CourseProgress, 
  CourseProgressSummary, 
  UpdateProgressPayload 
} from '../types/course';

export interface Course {
  id: number;
  title: string;
  slug: string;
  description: string;
  short_description: string;
  thumbnail_url: string;
  status: string | { value: string }; // Peut être une chaîne ou un objet avec une propriété value
  price: number;
  created_at: string;
  updated_at: string;
  instructor_id: number;
  instructor?: {
    id: number;
    first_name: string;
    last_name: string;
    email: string;
  };
  category_id: number | null;
  category?: {
    id: number;
    name: string;
    description?: string;
  };
  tags: Array<{ id: number; name: string }>;
  modules?: Array<{
    id: number;
    title: string;
    description: string | null;
    order: number;
  }>;
  // Champs supplémentaires pour la transformation
  image?: string;
  category_name?: string;
  instructor_name?: string;
  level?: string;
  duration?: number;
}

export interface Lesson {
  id: number;
  title: string;
  description: string;
  type: string;
  duration: number;
  order: number;
  content: string;
  video_url?: string;
  module_id: number;
}

export interface Module {
  id: number;
  title: string;
  description: string;
  order: number;
  lessons: Lesson[];
}

export interface CourseWithModules extends Omit<Course, 'modules'> {
  modules: Module[];
}

class CourseService {
  private static instance: CourseService;
  
  private constructor() {}

  public static getInstance(): CourseService {
    if (!CourseService.instance) {
      CourseService.instance = new CourseService();
    }
    return CourseService.instance;
  }

  public async getCourses(params?: {
    skip?: number;
    limit?: number;
    search?: string;
    level?: string;
    tag?: string;
  }): Promise<{ courses: Course[]; total: number }> {
    try {
      const response = await apiService.get<{ items: Course[]; total: number }>(
        API_CONFIG.ENDPOINTS.COURSES.BASE,
        { params }
      );
      return {
        courses: response.items,
        total: response.total,
      };
    } catch (error) {
      console.error('Erreur lors de la récupération des cours:', error);
      throw error;
    }
  }

  public async getCourseBySlug(slug: string): Promise<Course> {
    try {
      return await apiService.get<Course>(`${API_CONFIG.ENDPOINTS.COURSES.BASE}/slug/${slug}`);
    } catch (error) {
      console.error('Erreur lors de la récupération du cours:', error);
      throw error;
    }
  }

  public async getFeaturedCourses(): Promise<Course[]> {
    try {
      const response = await apiService.get<{ items: Course[] }>(
        API_CONFIG.ENDPOINTS.COURSES.FEATURED
      );
      return response.items;
    } catch (error) {
      console.error('Erreur lors de la récupération des cours à la une:', error);
      return [];
    }
  }

  public async searchCourses(query: string): Promise<Course[]> {
    try {
      const response = await apiService.get<{ items: Course[] }>(
        API_CONFIG.ENDPOINTS.COURSES.SEARCH,
        { params: { q: query } }
      );
      return response.items;
    } catch (error) {
      console.error('Erreur lors de la recherche de cours:', error);
      return [];
    }
  }

  public async enrollCourse(courseId: number): Promise<{ success: boolean }> {
    try {
      await apiService.post(`${API_CONFIG.ENDPOINTS.COURSES.BASE}/${courseId}/enroll`);
      return { success: true };
    } catch (error) {
      console.error('Erreur lors de l\'inscription au cours:', error);
      throw error;
    }
  }

  /**
   * Met à jour la progression d'une leçon pour l'utilisateur connecté
   * @param courseId ID du cours
   * @param lessonId ID de la leçon
   * @param isCompleted Si la leçon est marquée comme terminée
   * @param completionPercentage Pourcentage de complétion (0-100)
   */
  /**
   * Met à jour la progression d'une leçon pour l'utilisateur connecté
   * @param courseId ID du cours
   * @param lessonId ID de la leçon
   * @param payload Données de progression à mettre à jour
   */
  public async updateLessonProgress(
    courseId: number,
    lessonId: number,
    payload: UpdateProgressPayload
  ): Promise<{ success: boolean }> {
    try {
      await apiService.put(`/api/v1/progress/courses/${courseId}/lessons/${lessonId}/progress`, payload);
      return { success: true };
    } catch (error) {
      console.error('Erreur lors de la mise à jour de la progression:', error);
      throw error;
    }
  }

  /**
   * Récupère la progression d'un cours pour l'utilisateur connecté
   * @param courseId ID du cours
   */
  public async getCourseProgress(courseId: number): Promise<CourseProgress> {
    try {
      return await apiService.get(`/api/v1/progress/courses/${courseId}/progress`);
    } catch (error) {
      console.error('Erreur lors de la récupération de la progression du cours:', error);
      throw error;
    }
  }

  /**
   * Récupère un résumé de la progression de l'utilisateur dans tous ses cours
   */
  public async getUserProgressSummary(): Promise<CourseProgressSummary[]> {
    try {
      return await apiService.get('/api/v1/progress/my-progress');
    } catch (error) {
      console.error('Erreur lors de la récupération du résumé de progression:', error);
      return [];
    }
  }

  /**
   * Marque une leçon comme terminée
   */
  public async completeLesson(courseId: number, lessonId: number): Promise<{ success: boolean }> {
    return this.updateLessonProgress(courseId, lessonId, { 
      is_completed: true, 
      completion_percentage: 100 
    });
  }

  /**
   * Récupère les cours de l'enseignant
   */
  public async getTeacherCourses(): Promise<Course[]> {
    try {
      // L'endpoint pour récupérer les cours de l'enseignant est /courses/
      // Le préfixe /api/v1 est déjà inclus dans la configuration de base de l'API
      const response = await apiService.get<Course[]>('/courses/');
      return Array.isArray(response) ? response : [];
    } catch (error) {
      console.error('Error fetching teacher courses:', error);
      throw error;
    }
  }

  /**
   * Récupère tous les cours (admin uniquement)
   */
  public async getAdminCourses(params: {
    skip?: number;
    limit?: number;
  } = {}): Promise<{ courses: Course[]; total: number }> {
    try {
      const response = await apiService.get<Course[]>('/courses/admin/courses', { params });
      
      // Si la réponse est un tableau, c'est la liste des cours
      if (Array.isArray(response)) {
        return {
          courses: response,
          total: response.length
        };
      }
      
      // Si la réponse a une structure différente, essayer de l'adapter
      const courses = (response as any)?.data || [];
      const total = (response as any)?.total || courses.length;
      
      return {
        courses,
        total: typeof total === 'string' ? parseInt(total) : total
      };
    } catch (error) {
      console.error('Error fetching admin courses:', error);
      throw error;
    }
  }

  /**
   * Récupère les détails complets d'un cours avec ses modules et leçons
   */
  public async getCourseWithModules(courseId: number): Promise<CourseWithModules> {
    try {
      // 1. Récupérer les informations de base du cours
      const courseData = await apiService.get<any>(`teacher/dashboard/courses/${courseId}`);
      
      // 2. Récupérer les modules du cours
      const modulesData = await apiService.get<any>(`teacher/dashboard/courses/${courseId}/modules`);
      
      // 3. Pour chaque module, récupérer ses leçons
      const modules = [];
      for (const moduleData of Array.isArray(modulesData) ? modulesData : []) {
        try {
          const lessonsData = await apiService.get<any>(
            `teacher/dashboard/courses/${courseId}/modules/${moduleData.id}/lessons`
          );
          
          const lessons = Array.isArray(lessonsData) ? lessonsData.map((lesson: any) => ({
            id: lesson.id,
            title: lesson.title || 'Leçon sans titre',
            description: lesson.description || '',
            type: lesson.video_url ? 'video' : (lesson.type || 'text'),
            duration: lesson.duration_minutes || lesson.duration || 0,
            order: lesson.order_index || lesson.order || 0,
            content: lesson.content || '',
            video_url: lesson.video_url || '',
            module_id: lesson.module_id || moduleData.id
          })) : [];
          
          modules.push({
            id: moduleData.id,
            title: moduleData.title || 'Module sans titre',
            description: moduleData.description || '',
            order: moduleData.order_index || moduleData.order || 0,
            lessons
          });
        } catch (lessonError) {
          console.warn(`Impossible de récupérer les leçons du module ${moduleData.id}:`, lessonError);
          // Ajouter le module même sans leçons
          modules.push({
            id: moduleData.id,
            title: moduleData.title || 'Module sans titre',
            description: moduleData.description || '',
            order: moduleData.order_index || moduleData.order || 0,
            lessons: []
          });
        }
      }
      
      return {
        ...courseData,
        modules: modules.sort((a, b) => a.order - b.order)
      };
    } catch (error) {
      console.error('Erreur lors de la récupération du cours avec modules:', error);
      throw error;
    }
  }

  /**
   * Crée un nouveau cours pour l'enseignant
   */
  public async createCourse(courseData: {
    title: string;
    description: string;
    short_description?: string;
    category_id?: number;
    level?: string;
    price?: number;
    thumbnail_url?: string;
    status?: 'draft' | 'published';
  }): Promise<{ id: number; title: string; message: string }> {
    try {
      const response = await apiService.post<{ id: number; title: string; message: string }>(
        'courses/create',
        courseData
      );
      return response;
    } catch (error) {
      console.error('Erreur lors de la création du cours:', error);
      throw error;
    }
  }

  /**
   * Crée un nouveau module pour un cours
   */
  public async createModule(courseId: number, moduleData: {
    title: string;
    description?: string;
    order_index: number;
  }): Promise<{ id: number; title: string; message: string }> {
    try {
      const response = await apiService.post<{ id: number; title: string; message: string }>(
        `courses/${courseId}/modules`,
        moduleData
      );
      return response;
    } catch (error) {
      console.error('Erreur lors de la création du module:', error);
      throw error;
    }
  }

  /**
   * Crée une nouvelle leçon pour un module
   */
  public async createLesson(courseId: number, moduleId: number, lessonData: {
    title: string;
    description?: string;
    content?: string;
    video_url?: string;
    duration?: number;
    order_index: number;
    is_free?: boolean;
  }): Promise<{ id: number; title: string; message: string }> {
    try {
      const response = await apiService.post<{ id: number; title: string; message: string }>(
        `courses/${courseId}/modules/${moduleId}/lessons`,
        lessonData
      );
      return response;
    } catch (error) {
      console.error('Erreur lors de la création de la leçon:', error);
      throw error;
    }
  }

  /**
   * Crée un cours complet avec modules et leçons
   */
  public async createCourseWithContent(courseData: {
    // Informations de base du cours
    title: string;
    description: string;
    short_description?: string;
    category_id?: number;
    level?: string;
    price?: number;
    thumbnail_url?: string;
    status?: 'draft' | 'published';
    // Structure du cours
    modules: Array<{
      title: string;
      description?: string;
      lessons: Array<{
        title: string;
        description?: string;
        content?: string;
        video_url?: string;
        duration?: number;
        type?: 'video' | 'text' | 'quiz';
        is_free?: boolean;
      }>;
    }>;
  }): Promise<{ courseId: number; message: string }> {
    try {
      // 1. Créer le cours de base
      const courseResponse = await this.createCourse({
        title: courseData.title,
        description: courseData.description,
        short_description: courseData.short_description,
        category_id: courseData.category_id,
        level: courseData.level,
        price: courseData.price,
        thumbnail_url: courseData.thumbnail_url,
        status: courseData.status ?? 'draft'
      });

      const courseId = courseResponse.id;

      // 2. Créer les modules et leçons
      for (let moduleIndex = 0; moduleIndex < courseData.modules.length; moduleIndex++) {
        const moduleData = courseData.modules[moduleIndex];
        
        const moduleResponse = await this.createModule(courseId, {
          title: moduleData.title,
          description: moduleData.description,
          order_index: moduleIndex + 1
        });

        const moduleId = moduleResponse.id;

        // Créer les leçons du module
        for (let lessonIndex = 0; lessonIndex < moduleData.lessons.length; lessonIndex++) {
          const lessonData = moduleData.lessons[lessonIndex];
          
          await this.createLesson(courseId, moduleId, {
            title: lessonData.title,
            description: lessonData.description,
            content: lessonData.content,
            video_url: lessonData.video_url,
            duration: lessonData.duration,
            order_index: lessonIndex + 1,
            is_free: lessonData.is_free || false
          });
        }
      }

      return {
        courseId,
        message: `Cours "${courseData.title}" créé avec succès avec ${courseData.modules.length} modules`
      };
    } catch (error) {
      console.error('Erreur lors de la création du cours complet:', error);
      throw error;
    }
  }

  /**
   * Récupère les catégories disponibles pour la création de cours
   */
  public async getCategories(): Promise<Array<{ id: number; name: string; description?: string }>> {
    try {
      const response = await apiService.get<Array<{ id: number; name: string; description?: string }>>(
        'categories'
      );
      return Array.isArray(response) ? response : [];
    } catch (error) {
      console.error('Erreur lors de la récupération des catégories:', error);
      // Retourner des catégories par défaut en cas d'erreur
      return [
        { id: 1, name: 'Programmation', description: 'Cours de développement et programmation' },
        { id: 2, name: 'Design', description: 'Cours de design et créativité' },
        { id: 3, name: 'Business', description: 'Cours de business et entrepreneuriat' },
        { id: 4, name: 'Marketing', description: 'Cours de marketing et communication' },
        { id: 5, name: 'Général', description: 'Cours généraux' }
      ];
    }
  }

  /**
   * Récupère les cours auxquels l'étudiant est inscrit
   */
  public async getStudentEnrolledCourses(): Promise<CourseWithModules[]> {
    try {
      const response = await apiService.get<CourseWithModules[]>('courses/student/enrolled');
      return Array.isArray(response) ? response : [];
    } catch (error) {
      console.error('Erreur lors de la récupération des cours inscrits de l\'étudiant:', error);
      throw error;
    }
  }

  /**
   * Récupère la liste des cours auxquels l'étudiant n'est pas inscrit
   * @param params Paramètres de pagination et de filtrage
   * @returns Liste des cours non inscrits et nombre total
   */
  public async getUnenrolledCourses(params: {
    skip?: number;
    limit?: number;
  } = {}): Promise<{ courses: Course[]; total: number }> {
    try {
      const { skip = 0, limit = 10 } = params;
      
      console.log('Envoi de la requête pour les cours non inscrits avec les paramètres:', { skip, limit });
      
      // Utiliser POST avec les paramètres dans le corps
      const response = await apiService.post<Course[]>('/student/courses/unenrolled', {
        skip,
        limit
      });

      console.log('Réponse du backend (brute):', response);
      
      // Le backend renvoie directement un tableau de cours
      const courses = Array.isArray(response) 
        ? response.map(course => ({
            ...course,
            // S'assurer que le statut est une chaîne et non un objet CourseStatus
            status: course.status && typeof course.status === 'object' 
              ? course.status.value ?? 'published' 
              : String(course.status ?? 'published'),
            // Ajouter des valeurs par défaut pour les champs optionnels
            thumbnail_url: course.thumbnail_url ?? '',
            image: course.thumbnail_url ?? '',
            short_description: course.short_description ?? '',
            // Extraire le nom de la catégorie si c'est un objet
            category_name: course.category?.name ?? 'Non catégorisé',
            // Extraire le nom de l'instructeur si c'est un objet
            instructor_name: course.instructor 
              ? `${course.instructor.first_name ?? ''} ${course.instructor.last_name ?? ''}`.trim() || 'Instructeur inconnu'
              : 'Instructeur inconnu'
          }))
        : [];
      
      // Le total est la longueur de la liste des cours dans ce cas
      const total = courses.length;
      
      console.log('Cours non inscrits récupérés:', courses);
      console.log('Nombre total de cours non inscrits:', total);

      return {
        courses,
        total
      };
    } catch (error) {
      console.error('Erreur lors de la récupération des cours non inscrits:', error);
      throw error;
    }
  }

  /**
   * Inscrit un étudiant à un cours
   */
  public async enrollInCourse(courseId: number): Promise<{
    success: boolean;
    message: string;
    course_id: number;
    course_title: string;
  }> {
    try {
const response = await apiService.post<{
        success: boolean;
        message: string;
        course_id: number;
        course_title: string;
      }>(`/student/enroll/${courseId}`, {});
      
      console.log(`Inscription réussie au cours ${courseId}:`, response);
      return response;
    } catch (error) {
      console.error(`Erreur lors de l'inscription au cours ${courseId}:`, error);
      throw error;
    }
  }
}

export const courseService = CourseService.getInstance();
